src/hockeypuck/server/settings.go

```diff
@@ -242,36 +246,134 @@ func DefaultSettings() Settings {
 }
 
 func ParseSettings(data string) (*Settings, error) {
-	// Parse the configuration file as a template first
-	tmpl, err := template.New("config").Funcs(sprig.TxtFuncMap()).Funcs(envFuncMap()).Parse(data)
-	if err != nil {
-		return nil, errors.WithStack(err)
+	// Check if data contains template syntax - if so, process as template first
+	if strings.Contains(data, "{{") && strings.Contains(data, "}}") {
+		// Parse the configuration file as a template first
+		tmpl, err := template.New("config").Funcs(sprig.TxtFuncMap()).Funcs(envFuncMap()).Parse(data)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+
+		// Initialize a writer to render the template
+		w := &bytes.Buffer{}
+
+		// Render the template
+		err = tmpl.Execute(w, readEnv())
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+
+		data = w.String()
 	}
 
-	// Initialize a writer to render the template
-	w := &bytes.Buffer{}
+	// Fix TOML structure - move top-level keys that appear after sections to the beginning
+	data = restructureTOML(data)
 
-	// Render the template
-	err = tmpl.Execute(w, readEnv())
+	// Try parsing directly without wrapper first
+	settings := DefaultSettings()
+	_, err := toml.Decode(data, &settings)
 	if err != nil {
-		return nil, errors.WithStack(err)
+		// Try parsing with [hockeypuck] wrapper
+		var docWithWrapper struct {
+			Hockeypuck Settings `toml:"hockeypuck"`
+		}
+		docWithWrapper.Hockeypuck = DefaultSettings()
+		_, err = toml.Decode(data, &docWithWrapper)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		settings = docWithWrapper.Hockeypuck
 	}
 
-	var doc struct {
-		Hockeypuck Settings `toml:"hockeypuck"`
-	}
-	doc.Hockeypuck = DefaultSettings()
-	_, err = toml.Decode(w.String(), &doc)
+	err = settings.Conflux.Recon.Settings.Resolve()
 	if err != nil {
 		return nil, errors.WithStack(err)
 	}
 
-	err = doc.Hockeypuck.Conflux.Recon.Settings.Resolve()
-	if err != nil {
-		return nil, errors.WithStack(err)
+	return &settings, nil
+}
+
+// restructureTOML reorganizes TOML content to put top-level keys before sections
```

Comment:
What's the advantage of this? Is it so that it can write out its running config in a canonical form?

Reply:
There is a level of redundancy in existing TOML config

```
[hockeypuck]
key1=value1
key2=value2

[hockeypuck.recon...]
key3=value3
```
Re-factor lets you go with a simplified version
```
key1=value1

[recon...]
key3=value3

key2=value2
```
I put `key2` on purpose at the end to demonstrate that the function will reorder the config in it's canonnical form:
```
key1=value1
key2=value2

[recon...]
key3=value3
```

Reply:
Hm, but automatic rearrangement makes the toml file ambiguous. Say there's a key that's valid both at the root level and in the subsection - standard toml will handle them but rearranging the file will change the semantics. I think it's safer to reject toml that's been mangled, rather than try to guess what the user really meant...